src/main/scala/mesosphere/marathon/state/MarathonStore.scala

```diff
@@ -38,27 +39,35 @@ class MarathonStore[S <: MarathonState[_, S]](
       .recover(exceptionTransform(s"Could not fetch ${ct.runtimeClass.getSimpleName} with key: $key"))
   }
 
-  def modify(key: String)(f: Update): Future[S] = lockManager.executeSequentially(key) {
-    val res = store.load(prefix + key).flatMap {
-      case Some(entity) =>
-        bytesRead.update(entity.bytes.length)
-        val updated = f(() => stateFromBytes(entity.bytes.toArray))
-        val updatedEntity = entity.withNewContent(updated.toProtoByteArray)
-        bytesWritten.update(updatedEntity.bytes.length)
-        store.update(updatedEntity)
-      case None =>
-        val created = f(() => newState()).toProtoByteArray
-        bytesWritten.update(created.length)
-        store.create(prefix + key, created)
+  def modify(key: String, onSuccess: (S) => Unit = _ => ())(f: Update): Future[S] = {
+    lockManager.executeSequentially(key) {
+      log.debug(s"Modify $prefix$key")
+      val res = store.load(prefix + key).flatMap {
```

Comment:
Generally, this is incorrect.

As described in #2299, writes are not immediately visible. So instead of a fetch here, we should actually do a lookup in our cache. But this is probably an issue for another PR.